syntax = "proto3";

package paxos;

// Defines a VM service, which provides functionality for Proposer, Acceptor
// and Learner.
service VM {
	// Acceptors will receive these
	rpc handle_prepare (PrepareRequest) returns (PromiseResponse);
	rpc handle_accept (AcceptRequest) returns (OKResponse);

	// Learners will receive these
	rpc handle_learn (LearnRequest) returns (OKResponse);
}

message PrepareRequest {
	uint64 proposal_number = 1;
	uint64 decree_number = 2;
	string proposer = 3;
}

message AcceptRequest {
	uint64 proposal_number = 1;
	uint64 decree_number = 2;

	// This assumes that the value will always be an integer; this may not
	// be the case for our workloads. Revisit this.
	uint64 value = 3;

	string proposer = 4;
}

message PromiseResponse {
    // Whether we promise or not. Look at OneOf later.
    bool promised = 1;

	// Boolean that indicates whether there was any previously accepted
	// proposal for this decree
	bool had_previous = 2;

	// Proposal number and decree number. Notice that decree number is valid
	// even if had_previous is true.
	uint64 proposal_number = 3;
    uint64 decree_number = 4;

	// This assumes that the value will always be an integer; this may not
	// be the case for our workloads. Revisit this.
	uint64 value = 5;
	string acceptor = 6;
}

message LearnRequest {
	uint64 proposal_number = 1;
	uint64 decree_number = 2;

	// This assumes that the value will always be an integer; this may not
	// be the case for our workloads. Revisit this.
	uint64 value = 3;
	string acceptor = 4;
}

message OKResponse {
	bool response = 1;
}
